crates/esplora/src/blocking_ext.rs

```diff
@@ -68,250 +56,368 @@ pub trait EsploraExt {
+
+    // We want to have a corresponding checkpoint per height. We iterate the heights of anchors
+    // backwards, comparing it against our `local_tip`'s chain and our current set of
+    // `update_blocks` to see if a corresponding checkpoint already exists.
+    let anchor_heights = anchors
+        .iter()
+        .map(|(a, _)| a.anchor_block().height)
+        // filter out duplicate heights
+        .filter({
+            let mut prev_height = Option::<u32>::None;
+            move |h| match prev_height.replace(*h) {
+                None => true,
+                Some(prev_h) => prev_h != *h,
             }
+        })
+        // filter out heights that surpass the update tip
+        .filter(|h| *h <= update_tip_height)
+        .rev();
+
+    // We keep track of a checkpoint node of `local_tip` to make traversing the linked-list of
+    // checkpoints more efficient.
+    let mut curr_cp = local_tip.clone();
 
-            if let Some(last_active_index) = last_active_index {
-                last_active_indexes.insert(keychain, last_active_index);
+    for h in anchor_heights {
+        if let Some(cp) = curr_cp.query_from(h) {
+            curr_cp = cp.clone();
+            if cp.height() == h {
+                // blocks that already exist in checkpoint linked-list is also stored in
+                // `update_blocks` because we want to keep higher checkpoints of `local_chain`
+                update_blocks.insert(h, cp.hash());
+                continue;
             }
         }
-
-        Ok((graph, last_active_indexes))
+        if let btree_map::Entry::Vacant(entry) = update_blocks.entry(h) {
+            entry.insert(client.get_block_hash(h)?);
+        }
     }
 
-    fn sync(
-        &self,
-        misc_spks: impl IntoIterator<Item = ScriptBuf>,
-        txids: impl IntoIterator<Item = Txid>,
-        outpoints: impl IntoIterator<Item = OutPoint>,
-        parallel_requests: usize,
-    ) -> Result<TxGraph<ConfirmationTimeHeightAnchor>, Error> {
-        let mut graph = self
-            .full_scan(
-                [(
-                    (),
-                    misc_spks
-                        .into_iter()
-                        .enumerate()
-                        .map(|(i, spk)| (i as u32, spk)),
-                )]
-                .into(),
-                usize::MAX,
-                parallel_requests,
-            )
-            .map(|(g, _)| g)?;
-
-        let mut txids = txids.into_iter();
+    Ok(local_chain::Update {
+        tip: CheckPoint::from_block_ids(
+            update_blocks
+                .into_iter()
+                .map(|(height, hash)| BlockId { height, hash }),
+        )
+        .expect("must be in order"),
+        introduce_older_blocks: true,
+    })
+}
+
+/// This performs a full scan to get an update for the [`TxGraph`] and
+/// [`KeychainTxOutIndex`](bdk_chain::keychain::KeychainTxOutIndex).
+#[doc(hidden)]
+pub fn full_scan_for_index_and_graph_blocking<K: Ord + Clone>(
```

Comment:
why is this a `pub fn`

Reply:
So we can test it in the separate test folder. The other option is to move the tests in `src`.

Reply:
Yeah I'd move that test in here if it doesn't test the public API.